calyx/src/passes/well_formed.rs

```diff
@@ -208,6 +252,42 @@ impl Visitor for WellFormed {
                 }
             })?;
 
+        if let CellType::Component { name: id } = &cell.prototype {
+            let mut cellmap:LinkedHashMap<ir::Id, CellType> = LinkedHashMap::new();
+            for comp in comps.iter(){
+                if comp.name.eq(id) {
+                    for cell in comp.cells.iter() {
+                        if cell.borrow().is_external() {
+                            cellmap.insert(cell.clone_name(), cell.borrow().prototype.clone());
+                        }
+                    }
+                    break;
+                }
+            }
+            for (outcell, incell) in s.external_cells.iter() {
+                if let Some(t) = cellmap.get(&(*outcell)) {
+                    let proto = incell.borrow().prototype.clone();
+                    if !same_type(t, &proto) {
+                        return Err(Error::malformed_control
+                            ("type of fed-in external cell does not match original type"
+                        .to_string())
+                        .with_pos(incell.borrow().name()));
+                    }
+                    cellmap.remove(&(*outcell));
+                }
+                else {
+                    return Err(Error::malformed_control(format!("{} does not have external cell named {}",
+                    id,
+                    *outcell))
+                    .with_pos(outcell));
+                }
+            }
+            if cellmap.keys().len() != 0 {
+                return Err(Error::malformed_control(format!("unmentioned external cell: {}",
+            cellmap.keys().next().unwrap())).with_pos(s.comp.borrow().name()))
```

Comment:
Once you use the ContructVisitor approach, this will need to be immutable because you can’t change the cellmap in the shared data structure